x/gov/types.go

```diff
@@ -0,0 +1,82 @@
+package gov
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/stake"
+)
+
+var (
+	proposalTypes = []string{"Text", "ParameterChange", "SoftwareUpgrade"}
+)
+
+// Vote
+type Vote struct {
+	Voter      sdk.Address `json:"voter"`       //  address of the voter
+	ProposalID int64       `json:"proposal_id"` //  proposalID of the proposal
+	Option     string      `json:"option"`      //  option from OptionSet chosen by the voter
+}
+
+//-----------------------------------------------------------
+
+// Proposal
+type Proposal struct {
+	ProposalID   int64  `json:"proposal_id"`   //  ID of the proposal
+	Title        string `json:"title"`         //  Title of the proposal
+	Description  string `json:"description"`   //  Description of the proposal
+	ProposalType string `json:"proposal_type"` //  Type of proposal. Initial set {PlainTextProposal, SoftwareUpgradeProposal}
+
+	Status string `json:"string"` //  Status of the Proposal {Pending, Active, Passed, Rejected}
+
+	SubmitBlock  int64     `json:"submit_block"`  //  Height of the block where TxGovSubmitProposal was included
+	TotalDeposit sdk.Coins `json:"total_deposit"` //  Current deposit on this proposal. Initial value is set at InitialDeposit
+
+	VotingStartBlock int64 `json:"voting_start_block"` //  Height of the block where MinDeposit was reached. -1 if MinDeposit is not reached
+
+}
+
+// Procedure around Deposits for governance
+type DepositProcedure struct {
+	MinDeposit       sdk.Coins `json:"min_deposit"`        //  Minimum deposit for a proposal to enter voting period.
+	MaxDepositPeriod int64     `json:"max_deposit_period"` //  Maximum period for Atom holders to deposit on a proposal. Initial value: 2 months
+}
+
+// Procedure around Tallying votes in governance
+type TallyingProcedure struct {
+	Threshold         sdk.Rat `json:"threshold"`          //  Minimum propotion of Yes votes for proposal to pass. Initial value: 0.5
+	Veto              sdk.Rat `json:"veto"`               //  Minimum value of Veto votes to Total votes ratio for proposal to be vetoed. Initial value: 1/3
+	GovernancePenalty sdk.Rat `json:"governance_penalty"` //  Penalty if validator does not vote
+}
+
+// Procedure around Voting in governance
+type VotingProcedure struct {
+	VotingPeriod int64 `json:"voting_period"` //  Length of the voting period.
+}
+
+// List of valid proposal types
+type ProposalTypes []string
```

Comment:
proposal types should not be a string but a type of string with constant types: check out the `types/stake.go` bonded status for inspiration: https://github.com/cosmos/cosmos-sdk/blob/master/types/stake.go#L10-L17

of course it could be a type `string` instead of `byte` as I've done in my tests.

Reply:
also it's bad practice to have the word `Type` in the name of a type, try something alterntive such as `kind`

Reply:
Concurred with Rigel; @sunnya97 what was the reason for strings?

Reply:
(edit: see comment above)

Reply:
Not sure why having a special type helps.  I can turn it to a byte though.

Reply:
it helps because the type holds more explicit information as to what the object is to be used for.